### PR TITLE
Replicated all to all and test

### PR DESCRIPTION
**Pull request by lyprince**
Add support for replicated all to all, plus test of output, grads, and update

### CODE REVIEW

File: tests/test_collectives.py [hunk 1]
```diff
@@ -135,4 +140,4 @@ def test_collective(op: Callable[[torch.Tensor, int], torch.Tensor]) -> None:
 
 
 if __name__ == "__main__":
-    test_collective(pea.collectives.all_reduce_cross_replica_sum)
+    test_collective(pea.collectives.all_to_all_single_cross_replica)
```
Comment (DouglasOrr) [inline tests/test_collectives.py]: Seems a bit odd to have a `main` method in this test. Can this need be served by `./dev tests -k PATTERN`?

File: poptorch_experimental_addons/cpp/replicatedalltoall.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#include <popart/ir.hpp>
+#include <popart/op/collectives/collectives.hpp>
+#include <popart/opmanager.hpp>
+#include <popart/opserialiser.hpp>
+#include <popart/tensor.hpp>
+#pragma GCC diagnostic pop
+
+namespace popart {
+
+ReplicatedAllToAllOp::ReplicatedAllToAllOp(const OperatorIdentifier &_opid,
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/cpp/replicatedalltoall.cpp]: This looks like a different clang-format to the config we've got checked in—were/are you able to run it?

File: poptorch_experimental_addons/cpp/replicatedalltoallx.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    getInTensor(ReplicatedAllToAllOp::getInIndex()),
+    prog,
+    toGclCommGroup(op.getReplicaGrouping()), debugContext("replicatedAllToAll"),
+    allToAllOptions);
+
+  setOutTensor(ReplicatedAllToAllOp::getOutIndex(), output);
+}
+
+InputCreatorType
+ReplicatedAllToAllOpx::getInputCreatorType(InIndex) const {
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/cpp/replicatedalltoallx.cpp]: I think I need to learn about unwinding!

File: poptorch_experimental_addons/cpp/replicatedalltoall.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#include "replicatedalltoall.hpp"
+#include <algorithm>
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/cpp/replicatedalltoall.cpp]: I think this might be unused (can't see a `std::` from algorithm in this file).

**Review by DouglasOrr — COMMENTED** [review]
This looks good; a few things I don't fully understand, which I presume come from the popart collective ops design, but generally fine.

I was going to mention copyright headers, but CI will say that too.

Perhaps I might be tempted to inline all the `[ch].pp` into a single `.cpp` file, since no-one else depends on these headers, and it keeps everything together, even if the file gets a bit long. Thoughts?

File: tests/test_collectives.py (hunk 1 above)
Comment (lyprince) [reply tests/test_collectives.py]: Ah this was for me to debug tests. Can remove.

File: poptorch_experimental_addons/cpp/replicatedalltoall.cpp (hunk 2 above)
Comment (lyprince) [reply poptorch_experimental_addons/cpp/replicatedalltoall.cpp]: Just done so.

**Comment by lyprince** [comment]
I've inlined everything into `replicatedalltoall.cpp`. It's far from being the longest `.cpp` in this repo!

**Review by DouglasOrr — APPROVED** [review]
LGTM; thanks!

**Pull request merged**